### PR TITLE
[Named min timestamp leases] Wiring it all except legacy

### PR DESCRIPTION
**Pull request by ergo14**
This PR is all wiring

### CODE REVIEW

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 1]
```diff
@@ -173,20 +195,20 @@ public long currentTimeMillis() {
...
-        throw new UnsupportedOperationException("Not implemented yet!");
+        return timestampLeaseAcquirer.acquireNamedTimestampLeases(requests);
     }
 
     @Override
     public Map<TimestampLeaseName, Long> getMinLeasedTimestamps(Set<TimestampLeaseName> timestampNames) {
-        // TODO(aalouane): implement!
-        throw new UnsupportedOperationException("Not implemented yet!");
+        return minLeasedTimestampGetter.getMinLeasedTimestamps(timestampNames);
     }
```
Comment (ergo14) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:204]: Finishing this up this way, now I realize there is an alternative way for the unlocking:
- have the `TimestampLeaseAcquirer` return a `result|err` type where the `err` contains anything that needs to be unlocked
- do the unlocking in this class
It's possible for this to be annoying in that we may lose context/stacktrace, but we can pass any as needed

**Comment by ergo14** [comment]
The diff is not helpful given the nature of the refactor. It may be better to try and re-do the changes on your side and see if it matches and semantics are identical for existing things

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 2]
```diff
@@ -173,20 +195,20 @@ public long currentTimeMillis() {
...
 
     @Override
     public void close() {
         transactionStarter.close();
         commitTimestampGetter.close();
         lockLeaseService.close();
+        timestampLeaseAcquirer.close();
```
Comment (jkozlowski) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:211]: Can we FLUP later on wrapping this in some sort of closer. ANy exception thrown here (if someone does something wrong, will cause us to leak

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 3]
```diff
@@ -50,53 +52,73 @@ public final class RemoteTimelockServiceAdapter implements TimelockService, Auto
...
+            NamespacedConjureTimelockService conjureClient,
             LockLeaseService lockLeaseService,
-            TransactionStarter transactionStarter,
-            CommitTimestampGetter commitTimestampGetter) {
+            RequestBatchersFactory batcherFactory,
+            TimestampLeaseAcquirer timestampLeaseAcquirer,
+            MinLeasedTimestampGetter minLeasedTimestampGetter) {
         this.rpcClient = rpcClient;
-        this.conjureTimelockService = conjureTimelockService;
+        this.conjureTimelockService = conjureClient;
```
Comment (jkozlowski) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:66]: I'm reading all this as just rejigging, don't see anything concerning.

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java [hunk 4]
```diff
@@ -73,8 +76,9 @@ final class TimestampLeaseAcquirerImpl implements TimestampLeaseAcquirer {
         this.uuidSupplier = uuidSupplier;
     }
 
-    public TimestampLeaseAcquirerImpl(NamespacedTimestampLeaseService delegate, Unlocker unlocker) {
-        this(delegate, unlocker, UniqueIds::pseudoRandomUuidV4);
+    public static TimestampLeaseAcquirer create(NamespacedTimestampLeaseService delegate, LockTokenUnlocker unlocker) {
```
Comment (jkozlowski) [inline lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: Fine

File: timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/NamespacedClients.java [hunk 5]
```diff
@@ -71,11 +84,29 @@ static NamespacedClients from(String namespace, ProxyFactory proxyFactory) {
 
     @Value.Derived
     default TimelockService timelockService() {
+        LockTokenUnlocker unlocker = new LegacyLockTokenUnlocker(namespacedConjureTimelockService());
```
Comment (jkozlowski) [inline timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/NamespacedClients.java:87]: Test stuff, as long as tests pass I'm fine.

File: timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/NamespacedClients.java [hunk 6]
```diff
@@ -71,11 +84,29 @@ static NamespacedClients from(String namespace, ProxyFactory proxyFactory) {
...
+
+        LockLeaseService lockLeaseService = LockLeaseService.create(
```
Comment (jkozlowski) [inline timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/NamespacedClients.java:89]: Looks like we just moved this out from RemoteTimelockServiceAdapter, so it's not created there but passed as a dep.

**Review by jkozlowski — APPROVED** [review]
Added do-not-merge so this merges after previous PR

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java (hunk 2 above)
Comment (ergo14) [reply lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:211]: Tracked

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/DefaultLockAndTimestampServiceFactory.java [hunk 7]
```diff
@@ -383,6 +385,45 @@ private static LockAndTimestampServices getLockAndTimestampServices(
                 .build();
     }
 
+    private static RemoteTimelockServiceAdapter createRemoteTimelockServiceAdapter(
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/DefaultLockAndTimestampServiceFactory.java:388]: nit: this is not consistent with the other methods here that use get, but realistically let's just rename the others to create

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 8]
```diff
@@ -50,53 +52,73 @@ public final class RemoteTimelockServiceAdapter implements TimelockService, Auto
...
         this.conjureTimelockService = conjureTimelockService;
+        this.lockLeaseService = lockLeaseService;
+        this.transactionStarter = transactionStarter;
+        this.commitTimestampGetter = commitTimestampGetter;
+        this.timestampLeaseAcquirer = timestampLeaseAcquirer;
+        this.minLeasedTimestampGetter = minLeasedTimestampGetter;
     }
 
-    public static RemoteTimelockServiceAdapter create(
+    public static TimelockService create(
```
Comment (jeremyk-91) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:91]: Having both a static factory and a public constructor is pretty non-standard. When would we use the public constructor?

Also, can we accept this break? Large Internal Product and backup service use this.

**Review by jeremyk-91 — COMMENTED** [review]
~Small bits~

Actually I think we need to follow up on the underlying breaks here, they're nontrivial IMO

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java (hunk 8 above)
Comment (ergo14) [reply lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:91]: Two bits:
- static factory and public constructor: this was the case before, I shifted arguments as necessary for my change and pushed the private one to the top. I am happy to address the existing cruft in this PR.
- breaks: totally missed this, addressing in a separate PR soon

**Pull request merged**